### PR TITLE
Simplify `maybeOpenRxWindow` and improve docs

### PR DESCRIPTION
**Pull request by edsko**
@kazu-yamamoto This PR introduces the simplified algorithm @FinleyMcIlwaine alluded to in https://github.com/kazu-yamamoto/network-control/pull/5#issuecomment-2231281418 . I don't think this PR is critical; if you prefer the algorithm as-is, feel free to simply close this PR; but since you asked, here it is. The main idea is simply that we send a window update as soon as we have consumed more than half the window; after all, the goal of rate limiting is to ensure the peer cannot send us data faster than we can consume it.

I have also tried to explain in the documentation of `RxFlow` (my understanding of) what is happening. The diagram is very similar to what you had originally, but not quite the same; in particular, `rxfLimit` sits _within_ the window. See text for details.

I haven't had a chance yet to discuss this with @FinleyMcIlwaine (he is in a different time zone to me), so if you _do_ agree that this is a good please hold off on merging for just a bit, until he and I can discuss also.

### CODE REVIEW

**Comment by edsko** [comment]
I'll see if I can fix the tests (see also #6 ).

**Comment by edsko** [comment]
Fixed the tests (so this closes #6); I've changed the test approach a bit; see code.

I will also run a few benchmarks.

**Comment by edsko** [comment]
For one particular benchmark (`grapesy-kvstore --no-work-simulation`), I'm getting

`network-control` version | Percentage of frames that are `WINDOW_UPDATE`s 
-- | --
0.1.1 (prior to #5) | 19.70% 
0.1.2 (after #5) | 9.17%
after this PR (#7) | 5.87%

**Comment by edsko** [comment]
Ok, discussed this with @FinleyMcIlwaine , we think this PR is good to go. (I fixed two minor tihngs in the comments.)

File: Network/Control/Flow.hs [hunk 1]
```diff
@@ -81,25 +79,55 @@ txWindowSize TxFlow{..} = txfLimit - txfSent
 
 -- | Flow for receiving.
 --
+-- The goal of 'RxFlow' is to ensure that our network peer does not send us data
+-- faster than we can consume it. We therefore impose a buffer size (also known
+-- as a _window_) of unconsumed bytes that we are willing to receive from the
```
Comment (kazu-yamamoto) [inline Network/Control/Flow.hs]: In my understanding, *window* is the buffer size which is not filled by peer, not the total size of the buffer.
So, this explanation is a bit confusing.

File: Network/Control/Flow.hs [hunk 2]
```diff
@@ -163,25 +160,27 @@ maybeOpenRxWindow
...
   where
-    available = rxfLimit - rxfReceived
-    threshold = rxfBufSize `unsafeShiftR` 1
-    minSize = rxfBufSize `unsafeShiftR` 3
-    consumed' = rxfConsumed + consumed
-    rxfLimit' = consumed' + rxfBufSize
+    rxfConsumed' = rxfConsumed + consumed
+
+    -- Minimum window update size
+    threshold = rxfBufSize `div` 2
```
Comment (kazu-yamamoto) [inline Network/Control/Flow.hs]: Why don't you use `unsafeShiftR`?

**Review by kazu-yamamoto — CHANGES_REQUESTED** [review]
The algorithm is good.
Please resolve my comments.

File: Network/Control/Flow.hs (hunk 1 above)
Comment (edsko) [reply Network/Control/Flow.hs]: Have rewritten this as 

> The goal of 'RxFlow' is to ensure that our network peer does not send us data faster than we can consume it. We therefore impose a maximum number of unconsumed bytes that we are willing to receive from the peer, which we refer to as the buffer size:

File: Network/Control/Flow.hs (hunk 2 above)
Comment (edsko) [reply Network/Control/Flow.hs]: Changed this to be back at its original definition:

```haskell
    threshold = rxfBufSize `unsafeShiftR` 1
```

**Comment by edsko** [comment]
@kazu-yamamoto Thanks for your review; I've addressed your comments. Let me know if you'd like any other changes!

**Review by kazu-yamamoto — APPROVED** [review]
Ah, your definition of the buffer size is different from mine.
In my mind, I have a fixed-size buffer like C while you are assuming variable-length data as implemented in Haskell.
The current explanation makes sense to me.

**Comment by kazu-yamamoto** [comment]
Merged and released.
Thank you for your contribution!

**Pull request merged**